Modules/Forum/classes/class.ilObjForumNotificationDataProvider.php

```diff
@@ -258,12 +272,26 @@ private function readForumData() : void
 
             $row = $this->db->fetchAssoc($result);
 
+            $top_item_crs_ref = $this->tree->checkForParentType($this->getRefId(), 'crs');
```

Comment:
We should not execute both operations here. We should check for a `crs` parent first. If a `crs` could **not** be found, we should check for a `grp`.

I suggest to extract this caching into a new class or a trait, nesting `if` statements should be avoided. We should work with early returns here.